fv3core/stencils/d_sw.py

```diff
@@ -153,29 +146,67 @@ def heat_damping_term(ub, vb, gx, gy, rsin2, cosa_s, u2, v2, du2, dv2):
 
 
 @gtstencil()
-def heatdamping(
-    ub: sd,
-    vb: sd,
-    delp: sd,
-    fx: sd,
-    fy: sd,
-    gx: sd,
-    gy: sd,
-    rsin2: sd,
-    cosa_s: sd,
-    heat_source: sd,
-    diss_est: sd,
-    damp: float,
-    do_skeb: int,
+def heat_source_from_vorticity_damping(
+    ub: FloatField,
+    vb: FloatField,
+    ut: FloatField,
+    vt: FloatField,
+    u: FloatField,
+    v: FloatField,
+    delp: FloatField,
+    rsin2: FloatField,
+    cosa_s: FloatField,
+    rdx: FloatField,
+    rdy: FloatField,
+    heat_source: FloatField,
+    dissipation_estimate: FloatField,
+    kinetic_energy_fraction_to_damp: float,
+    calculate_dissipation_estimate: int,
 ):
+    """
+    Calculates heat source from vorticity damping implied by energy conservation.
+
+    Args:
+        ub (in)
+        vb (in)
+        ut (in)
+        vt (in)
+        u (in)
+        v (in)
+        delp (in)
+        rsin2 (in)
+        cosa_s (in)
+        rdx (in): radius of Earth multiplied by x-direction gridcell width
+        rdy (in): radius of Earth multiplied by y-direction gridcell width
+        heat_source (out): heat source from vorticity damping
+            implied by energy conservation
+        diss_est (out): dissipation estimate, only calculated if
+            calculate_dissipation_estimate is 1
+        kinetic_energy_fraction_to_damp (in): according to its comment in fv_arrays,
+            the fraction of kinetic energy to explicitly damp and convert into heat.
+            TODO: confirm this description is accurate, why is it multiplied
+            by 0.25 below?
+        calculate_dissipation_estimate (in): If 1, calculate dissipation estimate.
+            Equivalent in Fortran model is do_skeb
+    """
     with computation(PARALLEL), interval(...):
+        ubt = (ub + vt) * rdx
+        fy = u * rdx
+        gy = fy * ubt
+        vbt = (vb - ut) * rdy
+        fx = v * rdy
+        gx = fx * vbt
         u2 = fy + fy[0, 1, 0]
-        du2 = ub + ub[0, 1, 0]
+        du2 = ubt + ubt[0, 1, 0]
         v2 = fx + fx[1, 0, 0]
-        dv2 = vb + vb[1, 0, 0]
-        dampterm = heat_damping_term(ub, vb, gx, gy, rsin2, cosa_s, u2, v2, du2, dv2)
-        heat_source[0, 0, 0] = delp * (heat_source - damp * dampterm)
-        diss_est[0, 0, 0] = diss_est - dampterm if do_skeb == 1 else diss_est
+        dv2 = vbt + vbt[1, 0, 0]
+        dampterm = heat_damping_term(ubt, vbt, gx, gy, rsin2, cosa_s, u2, v2, du2, dv2)
+        heat_source[0, 0, 0] = delp * (
```

Comment:
Not a need, but at some point we should try to standardize whether to use the [0,0,0] indices on assignments or not

Reply:
Yeah I was leaning towards liking [0, 0, 0] but the code just looks so nice when there are few/no square brackets on either side. And for temporaries like dv2, you cannot put square brackets on the left.

Reply:
Also the square brackets bring an implication that you could do something other than [0, 0, 0], which you can't, but we've already discussed that. It does help remind you if you are looking at an input field or a temporary. Always tradeoffs.